### PR TITLE
Adding odo file index json to gitignore file

### LINKED ISSUE #2223: add odo-file-index.json entry into .gitignore
When a component is created `odo component create` the command should automatically add `./.odo/odo-file-index.json` to the `.gitignore` file.

If `.gitignore` doesn't exist it should be created even if the current directory is not a git repo as the user might be creating git repo later after the component is created.

### PR DESCRIPTION
**Pull request by prietyc123**
## What is the purpose of this change? What does it change?

Adding odo-file-index.json into .gitignore file
## Was the change discussed in an issue?
fixes #2223 

## How to test changes?

make test-cmd-push

### CODE REVIEW

**Comment by amitkrout** [comment]
@prietyc123 Fix the go formatting failure.
```
$ make validate
./scripts/check-gofmt.sh
gofmt ERROR - These files are not formated by gofmt:
./pkg/util/file_indexer.go
make: *** [gofmt] Error 1
The command "make validate" exited with 2.
```

File: pkg/util/file_indexer.go [hunk 1]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
...
+	return directory, nil
+}
+
+// Check .gitignore file exists or not
+// Add odo-file-index.json entry to .gitignore
+func checkGitignoreFile(directory string) {
+
+	ignoreFile, err := gitignoreFilePath(directory)
+	if err != nil {
+		log.Fatalf("file does not exist: %s", err)
```
Comment (cdrage) [inline pkg/util/file_indexer.go]: you shouldn't be using `log.Fatalf`, but instead be returning `err` in the function so that it errors up throughout the code until it his logging.

File: pkg/util/file_indexer.go [hunk 2]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
...
+	if err == nil {
+		file, err := os.OpenFile(ignoreFile, os.O_APPEND|os.O_RDWR, 0600)
+		if err != nil {
+			log.Fatalf("failed opening file: %s", err)
+		}
+		defer file.Close()
+
+		data, err := ioutil.ReadFile(ignoreFile)
+		if err != nil {
+			log.Panicf("failed reading data from file: %s", err)
```
Comment (cdrage) [inline pkg/util/file_indexer.go]: same here.. shouldn't be using log, get rid of all of it.

File: pkg/util/file_indexer.go [hunk 3]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
...
+func checkGitignoreFile(directory string) {
+
+	ignoreFile, err := gitignoreFilePath(directory)
+	if err != nil {
+		log.Fatalf("file does not exist: %s", err)
+	}
+
+	_, err = os.Stat(ignoreFile)
+
+	if err == nil {
```
Comment (cdrage) [inline pkg/util/file_indexer.go]: using `err == nill` and then putting code within here is unorthodox. why exactly are we continuing after erroring here: https://github.com/openshift/odo/pull/2305/files#diff-ded22a7f93f540164727d9088a72def4R118 ?

File: pkg/util/file_indexer.go [hunk 4]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
...
+		file, err := os.OpenFile(ignoreFile, os.O_APPEND|os.O_RDWR, 0600)
```
Comment (cdrage) [inline pkg/util/file_indexer.go]: have you tested this on windows?

File: pkg/util/file_indexer.go [hunk 5]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
...
+				log.Fatalf("failed opening file: %s", err)
+			}
+			defer file.Close()
+
+			if _, err = file.WriteString("\n" + ".odo/odo-file-index.json"); err != nil {
+				log.Panicf("failed writing string to file: %s", err)
+			}
+		}
+	} else {
+	}
```
Comment (cdrage) [inline pkg/util/file_indexer.go]: what

**Comment by prietyc123** [comment]
> Unfortunately there are too many nested if statements, specifically:
> 
> ```shell
> 	} else if os.IsNotExist(err) {
> 		files, err := ioutil.ReadDir(filepath.Join(directory))
> 		if err != nil {
> 			return errors.Wrap(err, "No such directory")
> 		}
> 		if len(files) > 1 {
> 			file, err := os.OpenFile(ignoreFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
> 			if err != nil {
> 				return errors.Wrap(err, "failed opening file")
> 			}
> 			defer file.Close()
> 
> 			if _, err := file.WriteString("\n" + ".odo/odo-file-index.json"); err != nil {
> 				return errors.Wrap(err, "failed writing string to file")
> 			}
> 		}
> ```
> 
> Which is unorthodox as we rely on erroring out of the function in order to continue.
> 
> Is there any way for you to do any peer programming with @amitkrout or @dharmit in the office? Go is pretty difficult to grasp especially with the way it handles errors :(

Hmm, Let me find out some better way to do this. Anyway I will discuss with @amitkrout or @dharmit whoever is available.

**Comment by prietyc123** [comment]
@cdrage Can you please do next round of review

File: pkg/util/file_indexer.go (hunk 3 above)
Comment (dharmit) [reply pkg/util/file_indexer.go]: This indeed looks unusual. In Go, we generally:
1. have a check for `if err != nil` and do something (mostly do some cleanup if required and throw the error back to caller or exit the program)
2. then, in the same function that has `if err != nil` check, we have more code that's supposed to be executed if there was no error.

In pseodo-code it might look something like below:
```go
func myAwesomeFunction(abc string) error {
    ignoreFile, err := gitIgnoreFilePath(directory)

    if err != nil {
        // do what needs to be done if there's an error
    }
    // now write the code that's supposed to be executed if there's no error in the step above.
}

File: pkg/util/file_indexer.go [hunk 6]
```diff
@@ -83,6 +85,75 @@ func resolveFilePath(directory string) (string, error) {
...
+		file, err := os.OpenFile(ignoreFile, os.O_APPEND|os.O_RDWR, 0600)
+		if err != nil {
+			return errors.Wrap(err, "failed opening file")
+		}
+		defer file.Close()
+
+		data, err := ioutil.ReadFile(ignoreFile)
+		if err != nil {
+			return errors.Wrap(err, "failed reading data from file")
+		} else {
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Like I mentioned in previous comment, in Go, we generally have a check for error and then the code that is supposed to be executed if there was no error. So in this case, we could simplify it to:
```go
data, err := ioutil.ReadFile(ignoreFile)
if err != nil {
	return errors.Wrap(err, "failed reading data from file")
}
// check whether .odo/odo-file-index.json is already in the .gitignore file
if !strings.Contains(string(data), ".odo/odo-file-index.json") {
	if _, err := file.WriteString("\n" + ".odo/odo-file-index.json"); err != nil {
			return errors.Wrap(err, "failed writing string to file")
	}
}
```

File: pkg/util/file_indexer.go [hunk 7]
```diff
@@ -83,6 +85,75 @@ func resolveFilePath(directory string) (string, error) {
...
+		}
+	} else if os.IsNotExist(err) {
+		files, err := ioutil.ReadDir(filepath.Join(directory))
+		if err != nil {
+			return errors.Wrap(err, "No such directory")
+		}
+		if len(files) > 1 {
+			file, err := os.OpenFile(ignoreFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+			if err != nil {
+				return errors.Wrap(err, "failed opening file")
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Which file are we talking about here?

File: pkg/util/file_indexer.go [hunk 8]
```diff
@@ -83,6 +85,75 @@ func resolveFilePath(directory string) (string, error) {
...
+			}
+			defer file.Close()
+
+			if _, err := file.WriteString("\n" + ".odo/odo-file-index.json"); err != nil {
+				return errors.Wrap(err, "failed writing string to file")
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Which string? Which file? We could rephrase this to something like `Failed to add odo-file-index.json to gitignore`. That's just a suggestion. You could, of course, change it to something else that conveys the user what *exactly* failed.

File: pkg/util/file_indexer.go (hunk 4 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: No not yet. Will test and update the result here.

File: pkg/util/file_indexer.go [hunk 9]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
 	return directory, nil
 }
 
+// gitignoreFilePath gives the filepath of the .gitignore file in the contex
```
[the remark this answers is not in the document]
Comment (prietyc123) [reply pkg/util/file_indexer.go]: Fixed the typo in the latest change.

File: pkg/util/file_indexer.go [hunk 10]
```diff
@@ -83,6 +86,77 @@ func resolveFilePath(directory string) (string, error) {
...
+func gitignoreFilePath(directory string) (string, error) {
```
[the remark this answers is not in the document]
Comment (prietyc123) [reply pkg/util/file_indexer.go]: Corrected the naming :-)

File: pkg/util/file_indexer.go (hunk 3 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: Thanks @dharmit for pseudo-code.
I have updated all the changes. Can you please review once again and let me know if any modification is required further.

File: pkg/util/file_indexer.go (hunk 8 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: For better understanding to the user I have done the modification in my latest change update.

File: pkg/util/file_indexer.go [hunk 11]
```diff
@@ -95,6 +151,12 @@ func RunIndexer(directory string, ignoreRules []string) (filesChanged []string,
 		return filesChanged, filesDeleted, err
 	}
 
+	// check for .gitignore file and add odo-file-index.json to .gitignore
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: I'm a bit sceptical about putting this here. This piece of code would run each time a user does `odo push`. Should this be a part of the push action or create, that is, `odo create`? I feel it should be in `odo create`. WDYT @kadel ?

File: pkg/util/file_indexer.go [hunk 12]
```diff
@@ -83,6 +85,60 @@ func resolveFilePath(directory string) (string, error) {
...
+		}
+	}
+	return nil
+}
+
+// Check .gitignore file exists or not
+func checkGitIgnoreFile(directory string) error {
+
+	ignoreFile, err := gitIgnoreFilePath(directory)
+	_, err = os.Stat(ignoreFile)
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Since we're doing `_, err = os.Stat(ignoreFile)`, I'm assuming that we're not interested in the `FileInfo` returned by `os.Stat`. In that case, why do we have a call to it here? I'm confused. Or am I looking at it the wrong way?

We're also not doing anything with the `err` here.

**Comment by amitkrout** [comment]
> /test v4.3-integration

wait till pr #2362 merge in

File: pkg/util/file_indexer.go [hunk 13]
```diff
@@ -83,6 +85,60 @@ func resolveFilePath(directory string) (string, error) {
...
+	files, _ := ioutil.ReadDir(filepath.Join(directory))
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Why are we ignoring the error returned by `ioutil.ReadDir`?

File: pkg/util/file_indexer.go [hunk 14]
```diff
@@ -83,6 +85,60 @@ func resolveFilePath(directory string) (string, error) {
...
+
+	if len(files) > 1 {
+		if err != nil {
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Which `err` this checking for? I think it's the one in the line `_, err = os.Stat(ignoreFile)`. What's the point in checking for that err here?

File: pkg/util/file_indexer.go [hunk 15]
```diff
@@ -83,6 +85,60 @@ func resolveFilePath(directory string) (string, error) {
...
+			}
+			file.Close()
+		}
+		err = addOdoFileIndex(ignoreFile)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: IIUC, `checkGitIgnoreFile` is doing multiple things:
1. Checks if the `.gitignore` file is present.
2. If it's not present, it creates the file.
3. Adds entry for `.odo/odo-file-index.json` in the `.gitignore` file.

I think we should separate the 3rd step into a separate function of its own. I get it that we've created a separate function `addOdoFileIndex` but calling it within `checkGitIgnoreFile` still defies the purpose of this function. We could add it to `file_indexer.go` after the call to `checkGitIgnoreFile`.

File: tests/integration/cmd_push_test.go [hunk 16]
```diff
@@ -298,6 +298,19 @@ var _ = Describe("odo push command tests", func() {
 		})
 	})
 
+	Context("when .gitignore file exists or not", func() {
+		It("should create and push the contents of a named component and include odo-file-index.json path to .gitignore file to exclude the contents", func() {
+			helper.CmdShouldPass("git", "clone", "https://github.com/openshift/nodejs-ex", context+"/nodejs-ex")
```
Comment (dharmit) [inline tests/integration/cmd_push_test.go]: Please use `helper.CopyExample` like in https://github.com/openshift/odo/blob/master/tests/integration/servicecatalog/cmd_link_unlink_test.go#L66.

File: pkg/util/file_indexer.go (hunk 11 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: @dharmit IMO this should be in ```odo push``` because ```odo-file-index.json``` will always creates after ```odo push``` and ofcourse the code operation comes into picture after creation of ```odo-file-index.json```. Lets wait how @kadel see it.

File: pkg/util/file_indexer.go (hunk 12 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: Yes here we are just checking the existence of ```.gitignore``` path and if it does not exists the we are using ```err ``` to verify and create .gitignore if doesn't.

File: pkg/util/file_indexer.go (hunk 13 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: if ```filepath.join(directory)``` exists it means it will already have .odo file so there is no need of error handling.

File: pkg/util/file_indexer.go (hunk 14 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: https://github.com/openshift/odo/pull/2305#discussion_r342974192

File: pkg/util/file_indexer.go (hunk 15 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: @kadel @girishramnani @mik-dass WDYT? Can you please address if you have any suggestion for this.

File: tests/integration/cmd_push_test.go (hunk 16 above)
Comment (prietyc123) [reply tests/integration/cmd_push_test.go]: I will add such new case in my test script.

File: pkg/util/file_indexer.go (hunk 15 above)
Comment (mik-dass) [reply pkg/util/file_indexer.go]: @prietyc123 I feel @dharmit is right here. This function should just check for the existence of the .gitignore file.Only step 1 should be performed here.

File: pkg/util/file_indexer.go (hunk 12 above)
Comment (kadel) [reply pkg/util/file_indexer.go]: the error checking should immediately follow `Stat` call. You should either move `if err!=nil` right after this, or move `os.Stat` down to where you are checking the error.

File: pkg/util/file_indexer.go (hunk 11 above)
Comment (kadel) [reply pkg/util/file_indexer.go]: both make sense. It depends when we are adding it to the `.gitignore`. If we add `odo-file-index.json` to `.gitignore` when component (localConfig) is created that all the function should be in `create` package. If we are adding it to `.gitignore` when the `odo-file-index.json` file is created for the first time it should be in the package that creates `odo-file-index.json` file

File: pkg/util/file_indexer.go (hunk 13 above)
Comment (kadel) [reply pkg/util/file_indexer.go]: better be safe ;-) If there is no need to explicitly ignore the error you should always check it.

File: pkg/util/file_indexer.go [hunk 17]
```diff
@@ -83,6 +85,62 @@ func resolveFilePath(directory string) (string, error) {
...
+}
+
+// Check .gitignore file exists or not
+func checkGitIgnoreFile(directory string) error {
+
+	ignoreFile, err := gitIgnoreFilePath(directory)
+	// err checks the existence of .gitignore and then creates if does not exists
+	_, err = os.Stat(ignoreFile)
+	// As the directory already exists it means it already has one file i.e .odo file in it so ignore err
```
Comment (kadel) [inline pkg/util/file_indexer.go]: This is not true. the directory can exist and be empty

File: pkg/util/file_indexer.go [hunk 18]
```diff
@@ -83,6 +85,62 @@ func resolveFilePath(directory string) (string, error) {
...
+	files, _ := ioutil.ReadDir(filepath.Join(directory))
```
Comment (kadel) [inline pkg/util/file_indexer.go]: Why do you need to know what files are in the directory?

File: pkg/util/file_indexer.go (hunk 15 above)
Comment (kadel) [reply pkg/util/file_indexer.go]: @dharmit +1

File: pkg/util/file_indexer.go (hunk 11 above)
Comment (kadel) [reply pkg/util/file_indexer.go]: It is really not necessary to run this more than once. And there is no harm in adding it to `.gitignore` even if the file doesn't exist yet. 
I think it makes sense to do in `odo create` when the component gets initialized and `config.yaml` is created.


We can have it in fileindex also, but than we have to make sure that it is getting called only once when the `odo-file-index.json` is being created, definitely not every time the file is updated.

File: tests/integration/cmd_push_test.go (hunk 16 above)
Comment (kadel) [reply tests/integration/cmd_push_test.go]: @prietyc123 this is just adding unnecessary test case and it is taking longer. Using just `CopyExample` is ok

**Comment by prietyc123** [comment]
Seems to be a platform issue
/retest

**Comment by amitkrout** [comment]
```
Submitted failure event to sentry (id=90bf0a60491a4877b897daa1c1ce62cf)
error: could not run steps: step root failed: could not wait for build: the build root failed with reason PushImageToRegistryFailed: Failed to push the image to the registry.
```
/retest

**Comment by amitkrout** [comment]
@prietyc123 can you please fix this
```
$ make golint
golangci-lint run ./... --timeout 5m
pkg/util/file_indexer.go:122:14: ineffectual assignment to `err` (ineffassign)
	ignoreFile, err := gitIgnoreFilePath(directory)
	            ^
make: *** [golint] Error 1
The command "make golint" exited with 2.
```

File: pkg/util/file_indexer.go (hunk 15 above)
Comment (prietyc123) [reply pkg/util/file_indexer.go]: @dharmit I have modified the content addressed by you :-)

**Comment by prietyc123** [comment]
> @prietyc123 can you please fix this
> 
> ```
> $ make golint
> golangci-lint run ./... --timeout 5m
> pkg/util/file_indexer.go:122:14: ineffectual assignment to `err` (ineffassign)
> 	ignoreFile, err := gitIgnoreFilePath(directory)
> 	            ^
> make: *** [golint] Error 1
> The command "make golint" exited with 2.
> ```

Yes. I have fixed it .

File: pkg/util/file_indexer.go [hunk 19]
```diff
@@ -83,6 +86,57 @@ func resolveFilePath(directory string) (string, error) {
...
+	}
+	defer file.Close()
+
+	if data, err = ioutil.ReadFile(ignoreFile); err != nil {
+		return errors.Wrap(err, "failed reading data from .gitignore file")
+	}
+	// check whether .odo/odo-file-index.json is already in the .gitignore file
+	if !strings.Contains(string(data), ".odo/odo-file-index.json") {
+		if _, err := file.WriteString("\n" + ".odo/odo-file-index.json"); err != nil {
+			return errors.Wrap(err, "failed to Add odo-file-index.json to .gitignore file")
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: Instead of hard-coding the `.odo/odo-file-index.json` value in above three lines, we could use the constants `fileIndexDirectory` & `fileIndexName` defined in this same file so that if the filename changes in future, we don't have to make changes at multiple places.

File: pkg/util/file_indexer.go [hunk 20]
```diff
@@ -96,20 +96,20 @@ func gitIgnoreFilePath(directory string) (string, error) {
...
-	if data, err = ioutil.ReadFile(ignoreFile); err != nil {
+	if data, err = ioutil.ReadFile(gitIgnoreFile); err != nil {
 		return errors.Wrap(err, "failed reading data from .gitignore file")
 	}
 	// check whether .odo/odo-file-index.json is already in the .gitignore file
-	if !strings.Contains(string(data), ".odo/odo-file-index.json") {
-		if _, err := file.WriteString("\n" + ".odo/odo-file-index.json"); err != nil {
+	if !strings.Contains(string(data), filepath.Join(fileIndexDirectory, fileIndexName)) {
+		if _, err := file.WriteString("\n" + filepath.Join(fileIndexDirectory, fileIndexName)); err != nil {
 			return errors.Wrap(err, "failed to Add odo-file-index.json to .gitignore file")
```
Comment (dharmit) [inline pkg/util/file_indexer.go]: You mean `"failed to add %v to .gitignore file", fileIndexName`? :smile:

Comment (prietyc123) [reply pkg/util/file_indexer.go]: Yes :-D

File: pkg/util/file_indexer.go [hunk 21]
```diff
@@ -83,6 +86,57 @@ func resolveFilePath(directory string) (string, error) {
...
+func checkGitIgnoreFile(directory string) (string, error) {
+
+	gitIgnoreFile, err := gitIgnoreFilePath(directory)
+	if err != nil {
+		return "", err
+	}
+	// err checks the existence of .gitignore and then creates if does not exists
+	_, err = os.Stat(gitIgnoreFile)
+
+	if err != nil {
```
Comment (mik-dass) [inline pkg/util/file_indexer.go]: The err can be other type too and thus may not indicate that the file doesn't exist. Please check the err to know the type like below

```
if _, err := os.Stat(filePath); os.IsNotExist(err) {
           // logic here
	}
```

**Comment by mik-dass** [comment]
@prietyc123 If possible, also maybe add some unit tests for those functions?

File: tests/integration/cmd_push_test.go [hunk 22]
```diff
@@ -318,7 +316,30 @@ var _ = Describe("odo push command tests", func() {
...
+
+	Context("when .gitignore file exists", func() {
+		It("should create and push the contents of a named component and include odo-file-index.json path to .gitignore file to exclude the contents", func() {
+			helper.CmdShouldPass("git", "clone", "https://github.com/openshift/nodejs-ex", context+"/nodejs-ex")
+			helper.CmdShouldPass("odo", "component", "create", "nodejs", cmpName, "--project", project, "--context", context+"/nodejs-ex", "--app", appName)
+
+			// push and include the odo-file-index.json path to .gitignore file
+			helper.CmdShouldPass("odo", "push", "--context", filepath.Join(context, "nodejs-ex"))
+			ignoreFilePath := filepath.Join(context, "nodejs-ex", ".gitignore")
+			helper.FileShouldContainSubstring(ignoreFilePath, filepath.Join(".odo", "odo-file-index.json"))
```
Comment (amitkrout) [inline tests/integration/cmd_push_test.go]: Then you need to verify inside the pod that ```odo-file-index.json``` is not pushed.

File: tests/integration/cmd_push_test.go [hunk 23]
```diff
@@ -318,7 +316,30 @@ var _ = Describe("odo push command tests", func() {
...
+
+	Context("when .gitignore file does not exist", func() {
+		It("should create and push the contents of a named component and also create .gitignore then include odo-file-index.json path to .gitignore file to exclude the contents", func() {
+			helper.CopyExample(filepath.Join("source", "nodejs"), context)
+			helper.CmdShouldPass("odo", "component", "create", "nodejs", cmpName, "--project", project, "--context", context, "--app", appName)
 
+			// push and include the odo-file-index.json path to .gitignore file
+			helper.CmdShouldPass("odo", "push", "--context", filepath.Join(context))
+			ignoreFilePath := filepath.Join(context, ".gitignore")
+			helper.FileShouldContainSubstring(ignoreFilePath, filepath.Join(".odo", "odo-file-index.json"))
```
Comment (amitkrout) [inline tests/integration/cmd_push_test.go]: https://github.com/openshift/odo/pull/2305/files#r346914510

File: tests/integration/cmd_push_test.go (hunk 22 above)
Comment (prietyc123) [reply tests/integration/cmd_push_test.go]: IMO as it is related to gitignore so while git push it should not be pushed. I think its nowhere related to pod and hence IMO it will not be the used case.

File: tests/integration/cmd_push_test.go (hunk 23 above)
Comment (prietyc123) [reply tests/integration/cmd_push_test.go]: https://github.com/openshift/odo/pull/2305#discussion_r347314598

**Comment by prietyc123** [comment]
#1940
/test v4.1-e2e-scenarios

File: tests/integration/cmd_push_test.go (hunk 22 above)
Comment (amitkrout) [reply tests/integration/cmd_push_test.go]: ok

**Comment by prietyc123** [comment]
> @prietyc123 If possible, also maybe add some unit tests for those functions?

https://github.com/openshift/odo/issues/2404

File: tests/integration/cmd_push_test.go (hunk 23 above)
Comment (amitkrout) [reply tests/integration/cmd_push_test.go]: /lgtm
